### PR TITLE
Fix ARIMA not accounting for gap in prediction from end of training data

### LINKED ISSUE #3853: `ARIMARegressor` does not support predictions of input with a gap off of training data
```
from evalml.pipelines.components import ARIMARegressor
from evalml.preprocessing import split_data
import pandas as pd

df = pd.read_csv("/Users/jeremy.shih/Downloads/souvenirs.csv")
df.ww.init()
y = df["Sales"]
X = df.drop('Sales', axis=1)

X_train, X_test, y_train, y_test = split_data(
        X, y, problem_type="time series regression", test_size=0.10, random_seed=0
    )

X_test_last_5 = X_test.tail(5)

arima = ARIMARegressor()
arima.fit(X_train, y_train)

assert (arima.predict(X_test).tail(5).tolist() == arima.predict(X_test_last_5).tolist())
```

Current `ARIMARegressor` creates a `ForecastingHorizon` with the `is_relative` flag turned on. This means that the ARIMA object will try to predict dates off of the last date known in the training data. Thus, it will provide the wrong predictions for dates with a gap off of the training data. 

A potential fix would be to use the `is_relative` flag but account for the difference in index between the last index in `X_train` and in `X_test`

```
def _set_forecast(self, X):
        from sktime.forecasting.base import ForecastingHorizon
        units_diff = X.index[0] - self.last_X_index
        fh_ = ForecastingHorizon([units_diff + i for i in range(len(X))], is_relative=True)
```

### PR DESCRIPTION
**Pull request by jeremyliweishih**
Fixes #3853.

### CODE REVIEW

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py [hunk 1]
```diff
@@ -142,7 +142,25 @@ def _match_indices(self, X, y):
     def _set_forecast(self, X):
         from sktime.forecasting.base import ForecastingHorizon
 
-        fh_ = ForecastingHorizon([i + 1 for i in range(len(X))], is_relative=True)
+        # we can only calculate the difference if the indices are of the same type
```
Comment (jeremyliweishih) [inline evalml/pipelines/components/estimators/regressors/arima_regressor.py:145]: set this as a fallback to previous behavior in case we ever receive inconsistent indices

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py [hunk 2]
```diff
@@ -142,7 +142,25 @@ def _match_indices(self, X, y):
...
+            units_diff = X.index[0] - self.last_X_index
+            if isinstance(X.index, pd.DatetimeIndex):
+                dates_diff = pd.date_range(
+                    start=self.last_X_index,
+                    end=X.index[0],
+                    freq=X.index.freq,
+                )
+                units_diff = len(dates_diff) - 1
+            fh_ = ForecastingHorizon(
+                [units_diff + i for i in range(len(X))],
```
Comment (eccabay) [inline evalml/pipelines/components/estimators/regressors/arima_regressor.py:156]: Seems like we needed an off-by-one offset in the non-gap case. Should that still be the case here as well?

File: evalml/tests/component_tests/test_arima_regressor.py [hunk 3]
```diff
@@ -139,6 +140,7 @@ def test_set_forecast(ts_data):
     )
 
     clf = ARIMARegressor()
+    clf.last_X_index = X.index[-1]
```
Comment (eccabay) [inline evalml/tests/component_tests/test_arima_regressor.py:143]: Why was this necessary?

File: evalml/tests/component_tests/test_arima_regressor.py [hunk 4]
```diff
@@ -397,3 +399,25 @@ def test_arima_regressor_respects_use_covariates(mock_predict, mock_fit, ts_data
     assert "y" in mock_fit.call_args.kwargs
     mock_predict.assert_called_once()
     assert "X" not in mock_predict.call_args.kwargs
+
+
+@pytest.mark.parametrize("use_covariates", [True, False])
+def test_arima_regressor_can_forecast_arbitrary_dates(use_covariates, ts_data):
```
Comment (eccabay) [inline evalml/tests/component_tests/test_arima_regressor.py:405]: It might be useful here to highlight how this is arbitrary dates in a comment or something - how long is X_test here, so how big of a gap is it between the training data and what we're asking to predict on?

File: evalml/tests/component_tests/test_arima_regressor.py [hunk 5]
```diff
@@ -397,3 +399,25 @@ def test_arima_regressor_respects_use_covariates(mock_predict, mock_fit, ts_data
...
+    )
+
+    X_test_last_5 = X_test.tail(5)
+
+    arima = ARIMARegressor(use_covariates=use_covariates)
+    arima.fit(X_train, y_train)
+
+    assert (
+        arima.predict(X_test).tail(5).tolist() == arima.predict(X_test_last_5).tolist()
+    )
```
Comment (eccabay) [inline evalml/tests/component_tests/test_arima_regressor.py:423]: Does this fail when run on current main?

**Review by eccabay — APPROVED** [review]
Looks pretty good, just had a few questions! Thanks for taking care of this.

File: evalml/tests/component_tests/test_arima_regressor.py (hunk 3 above)
Comment (jeremyliweishih) [reply evalml/tests/component_tests/test_arima_regressor.py:143]: we never ran fit in this test so `last_X_index` wasn't being set!

File: evalml/tests/component_tests/test_arima_regressor.py (hunk 4 above)
Comment (jeremyliweishih) [reply evalml/tests/component_tests/test_arima_regressor.py:405]: will add!

File: evalml/tests/component_tests/test_arima_regressor.py (hunk 5 above)
Comment (jeremyliweishih) [reply evalml/tests/component_tests/test_arima_regressor.py:423]: yes!

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py (hunk 1 above)
Comment (chukarsten) [reply evalml/pipelines/components/estimators/regressors/arima_regressor.py:145]: Is the index type being different indicative of something?

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py [hunk 6]
```diff
@@ -185,8 +203,10 @@ def fit(self, X, y=None):
 
         sp = self._get_sp(X)
         self._component_obj.sp = sp
+        self.last_X_index = X.index[-1] if X is not None else y.index
```
Comment (chukarsten) [inline evalml/pipelines/components/estimators/regressors/arima_regressor.py:206]: This looks kind of weird....either the value is a value of an index...or it's an index?

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py (hunk 2 above)
Comment (jeremyliweishih) [reply evalml/pipelines/components/estimators/regressors/arima_regressor.py:156]: We need the + 1 since `range(len(X))` starts at 0. In the gap case `units_diff` should always be > 0 so we don't need it!

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py [hunk 7]
```diff
@@ -226,7 +246,17 @@ def predict(self, X, y=None):
             },
         )
         if not X.empty and self.use_covariates:
-            y_pred = self._component_obj.predict(fh=fh_, X=X)
+            if fh_[0] != 1:
+                # pmdarima (which sktime uses under the hood) only forecasts off the training data
+                # but sktime circumvents this by predicting everything from the end of training data to the date / periods requested
+                # and only returning the values for dates / periods given to sktime. Because of this,
+                # pmdarima requires the number of covariate rows to equal the length of the total number of periods (X.shape[0] == fh_[-1]) if covariates are used.
+                # We circument this by adding arbitrary rows to the start of X since sktime discards these values when predicting.
```
Comment (chukarsten) [inline evalml/pipelines/components/estimators/regressors/arima_regressor.py:254]:
```suggestion
                # We circumvent this by adding arbitrary rows to the start of X since sktime discards these values when predicting.
```

**Review by chukarsten — APPROVED** [review]
Just some small things and questions.  Little nit to approve.

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py (hunk 1 above)
Comment (jeremyliweishih) [reply evalml/pipelines/components/estimators/regressors/arima_regressor.py:145]: it should only believe the result of inconsistent user behavior and this is a safeguard against that!

File: evalml/pipelines/components/estimators/regressors/arima_regressor.py (hunk 6 above)
Comment (jeremyliweishih) [reply evalml/pipelines/components/estimators/regressors/arima_regressor.py:206]: oops - will fix!
